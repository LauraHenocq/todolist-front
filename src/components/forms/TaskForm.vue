<script setup lang="ts">
import { computed, ref } from "vue";
import { TaskStatus, TaskEntity } from '@/models/task.entity';

type Props = {
  isEditionMode: Boolean,
  task?: TaskEntity | null
}

const { isEditionMode, task } = defineProps<Props>()

const isValid = ref(false);

const statusItems = [
  {
    label: 'A faire',
    value: TaskStatus.TODO,
  },
  {
    label: 'En cours',
    value: TaskStatus.IN_PROGRESS,
  },
  {
    label: 'Terminée',
    value: TaskStatus.DONE,
  }
]

const formData = ref({
  title: task && isEditionMode ? task.title : '',
  description: task &&  isEditionMode ? task.description : '',
  status: task && isEditionMode ? task.status : undefined
});

const emit = defineEmits(['create-task', 'update-task']);

const submitForm = () => {
  if (isValid.value) {
    if (isEditionMode) {
      emit('update-task', formData.value);
    } else {
      emit('create-task', { title: formData.value.title, description: formData.value.description });
    }

    formData.value = { title: '', description: '', status: task ? task.status : undefined };
  }
};

defineExpose({ submitForm });

const titleRules = computed(() => {
  return [
    value => {
      if (value) return true;

      return 'Le titre est requis.';
    },
    value => {
      if (value?.length <= 100) return true;

      return 'Le titre doit contenir moins de 100 caractères.';
    }
  ];
});

const descriptionRules = computed(() => {
  return [
    value => {
      if (value) return true;

      return 'La description est requise.';
    },
    value => {
      if (value?.length <= 500) return true;

      return 'La description doit contenir moins de 500 caractères.';
    }
  ];
});

</script>

<template>
  <v-form data-cy="task-form" v-model="isValid" ref="formRef" @submit.prevent>
    <v-container>
      <v-row>
        <v-col
          cols="12"
          md="4"
        >
          <v-text-field
            data-cy="task-form-title-field"
            v-model="formData.title"
            :rules="titleRules"
            label="Titre"
            required
          ></v-text-field>
        </v-col>
  
        <v-col
          cols="12"
          md="4"
        >
          <v-text-field
            data-cy="task-form-description-field"
            v-model="formData.description"
            :rules="descriptionRules"
            label="Description"
            required
          ></v-text-field>
        </v-col>

        <v-col
          v-if="isEditionMode"
          cols="12"
          md="4"
        >
        <v-select
          v-model="formData.status"
          label="Statut"
          item-title="label"
          item-value="value"
          :items="statusItems"
        ></v-select>
        </v-col>
      </v-row>
    </v-container>
  </v-form>
</template>