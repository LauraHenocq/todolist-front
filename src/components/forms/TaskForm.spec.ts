import { mount, VueWrapper } from '@vue/test-utils';
import { nextTick } from 'vue';
import { beforeEach, expect, it, describe } from 'vitest'
import { createVuetify } from 'vuetify';
import TaskForm from './TaskForm.vue';
import { TaskEntity, TaskStatus } from '@/models/task.entity';

const vuetify = createVuetify();

describe('TaskCard', () => {
  let wrapper: VueWrapper
  const task = new TaskEntity({
    title: 'Titre de la tâche',
    description: 'Description de la tâche',
    _status: TaskStatus.TODO,
    id: 'task-id'
  });

  const isEditionMode = false;

  const defaultProps = {
    isEditionMode,
    task
  }

  beforeEach(() => {
    wrapper = mount(TaskForm, {
      props: defaultProps,
      global: {
        plugins: [vuetify],
      },
    });
  });

  describe('render', () => {
    it('should render', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe('actions', () => {
    describe('Creation mode', () => {
      it('should emit create-task event when the task form is submitted', async () => {
        await wrapper.find('#input-0').setValue('Test Title');
        await nextTick();
        await wrapper.find('#input-2').setValue('Test Description');
        await nextTick();
        await wrapper.vm.submitForm();
        await nextTick();
      
        expect(wrapper.emitted('create-task')![0]).toEqual([{
          title: 'Test Title',
          description: 'Test Description'
        }]);
        expect(wrapper.emitted('create-task')).toHaveLength(1);
      });
    });

    describe('Edition mode', () => {
      it('should emit update-task event when the task form is submit', async () => {
        wrapper = mount(TaskForm, {
          props: {
            ...defaultProps,
            isEditionMode: true,
          },
          global: {
            plugins: [vuetify],
          },
        });

        await wrapper.find('#input-0').setValue('Test Title');
        await nextTick();
        await wrapper.find('#input-2').setValue('Test Description');
        await nextTick();
        await wrapper.find('#input-2').setValue('Test Description');
        await nextTick();
        await wrapper.vm.submitForm();
        await nextTick();
      
        expect(wrapper.emitted('update-task')![0]).toEqual([{
          title: 'Test Title',
          description: 'Test Description',
          status: 'TODO'
        }]);
        expect(wrapper.emitted('update-task')).toHaveLength(1);
      });
    });
  });
});