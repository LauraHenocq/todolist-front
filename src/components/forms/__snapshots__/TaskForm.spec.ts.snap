// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TaskCard > render > should render 1`] = `
"<form class="v-form" novalidate="" data-cy="task-form">
  <div class="v-container v-locale--is-ltr">
    <div class="v-row">
      <div class="v-col-md-4 v-col-12">
        <div class="v-input v-input--horizontal v-input--center-affix v-input--density-default v-theme--light v-locale--is-ltr v-text-field" data-cy="task-form-title-field">
          <!---->
          <div class="v-input__control">
            <div class="v-field v-field--center-affix v-field--variant-filled v-theme--light v-locale--is-ltr">
              <div class="v-field__overlay"></div>
              <div class="v-field__loader">
                <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px;" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
                  <!---->
                  <div class="v-progress-linear__background"></div>
                  <div class="v-progress-linear__buffer" style="width: 0%;"></div>
                  <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
                    <div class="v-progress-linear__indeterminate">
                      <div class="v-progress-linear__indeterminate long"></div>
                      <div class="v-progress-linear__indeterminate short"></div>
                    </div>
                  </transition-stub>
                  <!---->
                </div>
              </div>
              <!---->
              <div class="v-field__field" data-no-activator=""><label class="v-label v-field-label v-field-label--floating" aria-hidden="true" for="input-0">
                  <!---->Titre
                </label><label class="v-label v-field-label" for="input-0">
                  <!---->Titre
                </label>
                <!----><input size="1" type="text" id="input-0" aria-describedby="input-0-messages" required="" class="v-field__input" value="">
                <!---->
              </div>
              <!---->
              <!---->
              <div class="v-field__outline">
                <!---->
                <!---->
              </div>
            </div>
          </div>
          <!---->
          <div id="input-0-messages" class="v-input__details" role="alert" aria-live="polite">
            <transition-group-stub name="slide-y-transition" tag="div" appear="false" persisted="false" css="true" class="v-messages">
              <!---->
            </transition-group-stub>
            <!---->
          </div>
        </div>
      </div>
      <div class="v-col-md-4 v-col-12">
        <div class="v-input v-input--horizontal v-input--center-affix v-input--density-default v-theme--light v-locale--is-ltr v-text-field" data-cy="task-form-description-field">
          <!---->
          <div class="v-input__control">
            <div class="v-field v-field--center-affix v-field--variant-filled v-theme--light v-locale--is-ltr">
              <div class="v-field__overlay"></div>
              <div class="v-field__loader">
                <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px;" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
                  <!---->
                  <div class="v-progress-linear__background"></div>
                  <div class="v-progress-linear__buffer" style="width: 0%;"></div>
                  <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
                    <div class="v-progress-linear__indeterminate">
                      <div class="v-progress-linear__indeterminate long"></div>
                      <div class="v-progress-linear__indeterminate short"></div>
                    </div>
                  </transition-stub>
                  <!---->
                </div>
              </div>
              <!---->
              <div class="v-field__field" data-no-activator=""><label class="v-label v-field-label v-field-label--floating" aria-hidden="true" for="input-2">
                  <!---->Description
                </label><label class="v-label v-field-label" for="input-2">
                  <!---->Description
                </label>
                <!----><input size="1" type="text" id="input-2" aria-describedby="input-2-messages" required="" class="v-field__input" value="">
                <!---->
              </div>
              <!---->
              <!---->
              <div class="v-field__outline">
                <!---->
                <!---->
              </div>
            </div>
          </div>
          <!---->
          <div id="input-2-messages" class="v-input__details" role="alert" aria-live="polite">
            <transition-group-stub name="slide-y-transition" tag="div" appear="false" persisted="false" css="true" class="v-messages">
              <!---->
            </transition-group-stub>
            <!---->
          </div>
        </div>
      </div>
      <!--v-if-->
    </div>
  </div>
</form>"
`;
